cardano-testnet/test/Test/Cli/Babbage/LeadershipSchedule.hs

```diff
@@ -0,0 +1,194 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DisambiguateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+{- HLINT ignore "Redundant id" -}
+{- HLINT ignore "Redundant return" -}
+{- HLINT ignore "Use head" -}
+{- HLINT ignore "Use let" -}
+
+module Test.Cli.Babbage.LeadershipSchedule
+  ( hprop_leadershipSchedule
+  ) where
+
+import           Cardano.CLI.Shelley.Output (QueryTipLocalStateOutput (..))
+import           Control.Monad (void)
+import           Data.List ((\\))
+import           Data.Monoid (Last (..))
+import           GHC.Stack (callStack)
+import           Hedgehog (Property)
+import           Prelude
+import           System.Environment (getEnvironment)
+import           System.FilePath ((</>))
+
+import qualified Data.Aeson as J
+import qualified Data.Aeson.Types as J
+import qualified Data.List as L
+import qualified Data.Time.Clock as DTC
+import qualified Hedgehog as H
+import qualified Hedgehog.Extras.Stock.IO.Network.Sprocket as IO
+import qualified Hedgehog.Extras.Test.Base as H
+import qualified Hedgehog.Extras.Test.File as H
+import qualified Hedgehog.Extras.Test.Process as H
+import qualified System.Directory as IO
+import qualified System.Info as SYS
+import qualified Testnet.Util.Base as H
+
+import           Cardano.Testnet
+import           Testnet.Util.Assert
+import           Testnet.Util.Process
+import           Testnet.Util.Runtime
+
+hprop_leadershipSchedule :: Property
+hprop_leadershipSchedule = H.integrationRetryWorkspace 2 "babbage-leadership-schedule" $ \tempAbsBasePath' -> do
+  H.note_ SYS.os
+  base <- H.note =<< H.noteIO . IO.canonicalizePath =<< H.getProjectBase
+  configurationTemplate <- H.noteShow $ base </> "configuration/defaults/byron-mainnet/configuration.yaml"
+  conf@Conf { tempBaseAbsPath, tempAbsPath } <- H.noteShowM $
+    mkConf (ProjectBase base) (YamlFilePath configurationTemplate) tempAbsBasePath' Nothing
+
+  work <- H.note $ tempAbsPath </> "work"
+  H.createDirectoryIfMissing work
+
+  let
+    testnetOptions = BabbageOnlyTestnetOptions $ babbageDefaultTestnetOptions
+      { babbageNodeLoggingFormat = NodeLoggingFormatAsJson
+      }
+  tr@TestnetRuntime
+    { testnetMagic
+    , poolNodes
+    -- , wallets
+    -- , delegators
+    } <- testnet testnetOptions conf
+
+  poolNode1 <- H.headM poolNodes
```

Comment:
Instead of using `head` we use `headM`.  This is important because `head` is a partial function that can fail with an exception, which makes `head` not exception safe.